packages/commonwealth/client/scripts/state/api/comments/createComment.ts

```diff
@@ -0,0 +1,100 @@
+import { useMutation, useQueryClient } from '@tanstack/react-query';
+import axios from 'axios';
+import { updateLastVisited } from 'controllers/app/login';
+import Comment from 'models/Comment';
+import Thread from 'models/Thread';
+import app from 'state';
+import { ApiEndpoints } from 'state/api/config';
+import useFetchCommentsQuery from './fetchComments';
+
+interface CreateCommentProps {
+  address: string;
+  threadId: number,
+  chainId: string;
+  unescapedText: string,
+  parentCommentId: any,
+  attachments?: string[]
+}
+
+const createComment = async ({
+  chainId,
+  address,
+  threadId,
+  attachments,
+  unescapedText,
+  parentCommentId = null,
+}: CreateCommentProps) => {
+  const {
+    session = null,
+    action = null,
+    hash = null,
+  } = await app.sessions.signComment({
+    thread_id: threadId,
+    body: unescapedText,
+    parent_comment_id: parentCommentId,
+  });
+
+  const response = await axios.post(
+    `${app.serverUrl()}/threads/${threadId}/comments`,
+    {
+      author_chain: chainId,
+      chain: chainId,
+      address: address,
+      parent_id: parentCommentId,
+      // chain_entity_id: chainEntity?.id,
+      'attachments[]': attachments,
+      text: encodeURIComponent(unescapedText),
+      jwt: app.user.jwt,
+      canvas_action: action,
+      canvas_session: session,
+      canvas_hash: hash,
+    }
+  );
+
+  return new Comment(response.data.result)
+};
+
+const useCreateCommentMutation = ({ chainId, threadId }: Partial<CreateCommentProps>) => {
+  const queryClient = useQueryClient();
+  const { data: comments } = useFetchCommentsQuery({
+    chainId,
+    threadId,
+  })
+
+  return useMutation({
+    mutationFn: createComment,
+    onSuccess: async (newComment) => {
+      // update fetch comments query state
+      const key = [ApiEndpoints.FETCH_COMMENTS, chainId, threadId]
+      queryClient.cancelQueries({ queryKey: key });
+      queryClient.setQueryData([...key],
+        () => {
+          const updatedComments = [...(comments || []).filter(x => x.id !== newComment.id), newComment]
+          return updatedComments
+        }
+      );
+
+      // TODO: these types of async calls should also have a dedicated react query handler
+      // update last visisted
+      updateLastVisited(app.chain.meta, true)
```

Comment:
> It only gets triggered during the thread and comment creation which is weird for the lastVisited field

Ah, I had a similar thought during dev. I think @rbennettcw / @jnaviask might have some context on if we need the `updateLastVisited()` util which calls the `/writeUserSetting` api.

Reply:
Currently, the server already updates the user address' `lastActive` state when the user takes an action.

This `updateLastVisited` function almost appears to be duplicate functionality. The difference is that the `/writeUserSetting` route updates `User.lastVisited` while the existing logic updates the user's `Address.lastActive`.

Technically, we can derive the user's latest activity from their addresses… so this can be removed I think.

@jnaviask Thoughts on this?

Reply:
Looks like `lastVisited` is used in the /status route, which means we should be a bit careful about updating the field, because it determines how data is loaded for the user. In this case, it looks like the lastActive data is used to query "unseen posts" for the user (specifically: their last interaction in a community = the time we mark anything after as being "unseen").

@rbennettcw take a look at the code in /status and see if there's any other way we can replicate this query, as it's the _only_ place in the codebase we use lastVisited. Otherwise we may have to leave it.

Reply:
@jnaviask I took a quick look. It'll take some investigation, so I've created a new ticket for it: https://github.com/hicommonwealth/commonwealth/issues/4576